overlord/devicestate/devicestate_test.go

```diff
@@ -1166,6 +1167,192 @@ func (s *deviceMgrSuite) TestDeviceManagerEmptySystemModeRun(c *C) {
+	c.Assert(err, IsNil)
+	c.Check(smi, DeepEquals, &devicestate.SystemModeInfo{
+		Mode:   "run",
+		Seeded: false,
+	})
+
+	// seeded
+	s.state.Set("seeded", true)
+
+	smi, err = mgr.SystemModeInfo()
+	c.Assert(err, IsNil)
+	c.Check(smi, DeepEquals, &devicestate.SystemModeInfo{
+		Mode:   "run",
+		Seeded: true,
+	})
+}
+
+func (s *deviceMgrSuite) TestDeviceManagerSystemModeInfoUC20Install(c *C) {
+	modeEnv := &boot.Modeenv{Mode: "install"}
+	err := modeEnv.WriteTo("")
+	c.Assert(err, IsNil)
+
+	runner := s.o.TaskRunner()
+	mgr, err := devicestate.Manager(s.state, s.hookMgr, runner, s.newStore)
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// have a model
+	s.makeModelAssertionInState(c, "canonical", "pc-20", map[string]interface{}{
+		"architecture": "amd64",
+		// UC20
+		"grade": "dangerous",
+		"base":  "core20",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              snaptest.AssertedSnapID("oc-kernel"),
+				"type":            "kernel",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              snaptest.AssertedSnapID("pc"),
+				"type":            "gadget",
+				"default-channel": "20",
+			},
+		},
+	})
+
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand: "canonical",
+		Model: "pc-20",
+	})
+
+	// seeded
+	s.state.Set("seeded", true)
+	// no flags
+	c.Assert(boot.InitramfsExposeBootFlagsForSystem(nil), IsNil)
+	// data present
+	ubuntuData := filepath.Dir(boot.InstallHostWritableDir)
+	c.Assert(os.MkdirAll(ubuntuData, 0755), IsNil)
+
+	smi, err := mgr.SystemModeInfo()
+	c.Assert(err, IsNil)
+	c.Check(smi, DeepEquals, &devicestate.SystemModeInfo{
+		Mode:              "install",
+		HasModeenv:        true,
+		Seeded:            true,
+		BootFlags:         []string{},
+		HostDataLocations: []string{ubuntuData},
+	})
+
+	// factory
+	c.Assert(boot.InitramfsExposeBootFlagsForSystem([]string{"factory"}), IsNil)
+	smi, err = mgr.SystemModeInfo()
+	c.Assert(err, IsNil)
+	c.Check(smi, DeepEquals, &devicestate.SystemModeInfo{
+		Mode:              "install",
+		HasModeenv:        true,
+		Seeded:            true,
+		BootFlags:         []string{"factory"},
+		HostDataLocations: []string{ubuntuData},
+	})
+}
+
+func (s *deviceMgrSuite) TestDeviceManagerSystemModeInfoUC20Run(c *C) {
+	modeEnv := &boot.Modeenv{Mode: "run"}
+	err := modeEnv.WriteTo("")
+	c.Assert(err, IsNil)
+
+	runner := s.o.TaskRunner()
+	mgr, err := devicestate.Manager(s.state, s.hookMgr, runner, s.newStore)
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// have a model
+	s.makeModelAssertionInState(c, "canonical", "pc-20", map[string]interface{}{
+		"architecture": "amd64",
+		// UC20
+		"grade": "dangerous",
+		"base":  "core20",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              snaptest.AssertedSnapID("oc-kernel"),
```

Comment:
Is this a typo here? I.e. `pc-kernel` vs "**o**c-kernel" ? Not that it matters, just wondering :)

Reply:
interesting, I copied this from somewhere, so maybe we have the time around already, will look